File: 45-jump-game-ii/jump-game-ii.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        
        hMap = {0:[0]} # Holds the steps until this index


        for i,n in enumerate(nums):
            if len(nums) == 1: return 0
            stepTillThis = min(hMap[i]) if hMap[i] else float('inf')
            # print(i,hMap)
            # if (i == len(nums)-1):
            #     return min(hMap[i])

            for j in range(n):
                ind = i + j + 1 # looked index to add steps

                if (ind) in hMap:
                    hMap[ind].append(stepTillThis+1)
                else:
                    hMap[ind] = [stepTillThis+1]
                if (ind == len(nums)-1):
                    # print(ind, hMap[ind])
                    return stepTillThis+1
                    # return min(hMap[ind])

        # print(hMap)

        # return 0




'''
[2,3,1,1,4]
   1 1
     2 2 2
       2 3

her satir kac farkli yoldan gelinebilecegini ve kac adimda gelinecegini gosterir           

'''

File: 45-jump-game-ii/jump-game-ii.dart
class Solution {
  int jump(List<int> nums) {
      int p = 0;
      List<int> jumps = List.filled(nums.length, 0);

      while(p < nums.length){
          for(int i = 1; i <= nums[p]; i++ ){
              int scanner = p + i;
              if (scanner >= nums.length){
                  break;
              }
              if (jumps[scanner] == 0){
                  jumps[scanner] = jumps[p]+1;
              }

          }

          p++;
      }
      return jumps.last;

    
  }
}


//  0 1 1 2 2 3 4
// [2,3,1,0,1,2,4]
//            |